msteams-platform/samples/virtual-assistant.md

```diff
@@ -0,0 +1,338 @@
+
+```csharp
+    public static class ActivityExtensions
+    {
+        // Fetches skillId from CardAction data if present
+        public static string GetSkillId(this Activity activity)
+        {
+            string skillId = string.Empty;
+
+            try
+            {
+                if (activity.Type.Equals(ActivityTypes.Message) && activity.Value != null)
+                {
+                    var data = JsonConvert.DeserializeObject<SkillCardActionData>(activity.Value.ToString());
+                    skillId = data.SkillId;
+                }
+                else if (activity.Type.Equals(ActivityTypes.Invoke) && activity.Value != null)
+                {
+                    var data = JsonConvert.DeserializeObject<SkillCardActionData>(JObject.Parse(activity.Value.ToString()).SelectToken("data").ToString());
+                    skillId = data.SkillId;
+                }
+            }
+            catch
+            {
+                // If not able to retrive skillId, empty skillId should be returned
+            }
+
+            return skillId;
+        }
+    }
+```
+
+### Handle interruptions gracefully
+
+Virtual Assistant can handle interruptions in cases where a user tries to invoke a skill while another skill is currently active. we have introduced `TeamsSkillDialog` and `TeamsSwitchSkillDialog`, based on Bot Framework's [SkillDialog](https://github.com/microsoft/botframework-solutions/blob/5b46d73e220bbb4fba86c48be532e495535ca78a/sdk/csharp/libraries/microsoft.bot.solutions/Skills/SkillDialog.cs) and [SwitchSkillDialog](https://github.com/microsoft/botframework-solutions/blob/6d40fa8ae05f96b0c5e0464e01361a9e1deb696c/sdk/csharp/libraries/microsoft.bot.solutions/Skills/Dialogs/SwitchSkillDialog.cs), to enable users to switch a skill experience from card actions. To handle this request the Virtual Assistant prompts the user with a confirmation message to switch skills.
+
+![Confirmation prompt when switching to a new skill](../assets/images/bots/virtual-assistant/switch-skills-prompt.png)
+
+### Handling task module requests
+
+To add task module capabilities to a Virtual Assistant, two additional methods are included in the Virtual Assistant activity handler: `OnTeamsTaskModuleFetchAsync` and `OnTeamsTaskModuleSubmitAsync`. These methods listen to task module-related activities from Virtual Assistant, identify the skill associated with the request, and forward the request to the identified skill. 
+
+Request forwarding is done via the  [SkillHttpClient](/dotnet/api/microsoft.bot.builder.integration.aspnet.core.skills.skillhttpclient?view=botbuilder-dotnet-stable), `PostActivityAsync` method. It returns the response as `InvokeResponse` which is parsed and converted to `TaskModuleResponse` and sent back to Virtual Assistant.
```

Comment:
"It returns the response as `InvokeResponse` which is parsed and converted to `TaskModuleResponse`." instead of "It returns the response as `InvokeResponse` which is parsed and converted to `TaskModuleResponse` and sent back to Virtual Assistant." ?